src/sentry/tasks/assemble.py

```diff
@@ -455,6 +459,45 @@ def post_assemble(self):
         with metrics.timer("tasks.assemble.artifact_bundle"):
             self._create_artifact_bundle()
 
+        # We want to run the new indexing together with the old one, just to measure its performance.
+        if features.has("organizations:sourcemaps-bundle-flat-file-indexing", self.organization):
+            with metrics.timer("tasks.assemble.artifact_bundle_new"):
+                self._create_artifact_bundle_new()
+
+    def _create_artifact_bundle_new(self):
+        # We want to give precedence to the request fields and only if they are unset fallback to the manifest's
+        # contents.
+        self.release = self.release or self.archive.manifest.get("release")
+        self.dist = self.dist or self.archive.manifest.get("dist")
+
+        # We don't allow an upload without debug ids and release.
+        if not self.archive.has_debug_ids() and not self.release:
+            raise AssembleArtifactsError(
+                "uploading a bundle without debug ids or release is prohibited"
+            )
+
+        # We take a snapshot in time in order to have consistent values in the database.
+        date_snapshot = timezone.now()
+
+        bundle_id = self.archive.extract_bundle_id()
+
+        # We want to build all the required parameters for indexing.
+        artifact_bundle, created = self._bind_or_create_artifact_bundle(
+            bundle_id=bundle_id, date_added=date_snapshot
+        )
```

Comment:
I would prefer to keep the paths mutually exclusive, I just missed the else.

Reply:
Well, now you have duplicated about ~30 loc for no reason, and you are missing all the code that creates the `ReleaseArtifactBundle` etc associations.

You want to keep the old code functional, and *optionally* create a flat file index for the uploaded bundle based on feature flags.

Reply:
The thing is that we need to clarify whether we want to run them close to each other, I was thinking that we could enable the flag only for a subset of customers and not run the old system.

Reply:
Code duplication is also on purpose, to make the two completely independent from the starting root call and easier to refactor in case we decide to remove the old code.

Reply:
Lets run the old system in parallel as long as that works.

And this is not only about just the DB-based index vs the file based index. This is also about release association, debug-ids in the DB, which I believe you want to have regardless, etc.

Reply:
I am not sure we really want the associations in the db.

The main argument for having both in db and flat index is for querying performance on the endpoints for the UI. This is because the UI has a very different querying logic than the lookup endpoint (the lookup only cares about a specific bundle whereas the UI cares about multiple bundles).

Reply:
Not sure how I can make this clearer? For me this point is non-negotiable, for the following reasons:

At the very least, the `Project` association **has to be there**, as otherwise you (aka symbolicator) won’t be able to download the artifact bundle at all, as the permissions are tied to the project right now. Ever since https://github.com/getsentry/sentry/pull/49680, you can only download artifact bundles for the `Project` matching the URL for security reasons.

And as you mentioned, the association also drives the UI, so it means if there is no association anywhere, then customers will upload stuff that suddenly stops showing up in the UI and then they will swamp customer support with questions.

Plus, this point is also about backwards compatibility. We want to index into flat files based on the feature flag, while we are still indexing the way we currently do, and we know that it currently works.